src/EncodingContext.h

```diff
@@ -2,117 +2,49 @@
-        {
-        public:
-            typedef EncoderContextThriftIterator self;
-            typedef ContextElement* pointer;
-            typedef ContextElement& reference;
-
-            EncoderContextThriftIterator(const std::list<ContextElement>::iterator& it);
-            ~EncoderContextThriftIterator();
-            reference operator*() const;
-            pointer operator->() const;
-            self& operator++(int);
-            bool operator==(const self& obj) const;
-            bool operator!=(const self& obj) const;
-
-
-        private:
-            std::list<ContextElement>::iterator m_it;
-        };
 
+        class ContextElement {
 
-        class EncoderContextThriftConstIterator
-        {
         public:
-            typedef EncoderContextThriftConstIterator self;
-            typedef const ContextElement* pointer;
-            typedef const ContextElement& reference;
-
-            EncoderContextThriftConstIterator(const std::list<ContextElement>::const_iterator& it);
-            ~EncoderContextThriftConstIterator();
-            reference operator*() const;
-            pointer operator->() const;
-            self& operator++(int);
-            bool operator==(const self& obj) const;
-            bool operator!=(const self& obj) const;
 
+            typedef const Span& reference;
+            ContextElement(Span* span, const Encoder& encoder);
+            std::string ToString() const;
+            reference operator*() const { return *m_span; }
 
         private:
-            std::list<ContextElement>::const_iterator m_it;
+            std::unique_ptr<Span> m_span;
+            const Encoder&        m_encoder;
         };
 
-    public:
-        typedef EncoderContextThriftIterator iterator;
-        typedef EncoderContextThriftConstIterator const_iterator;
+        typedef std::vector<ContextElement> ContextElementVector;
+        typedef ContextElementVector::iterator iterator;
+        typedef ContextElementVector::const_iterator const_iterator;
 
-        EncoderContextThrift();
-        EncoderContextThrift(const char* buf, uint32_t sz);
-        virtual ~EncoderContextThrift();
+        EncoderContext();
+        EncoderContext(const std::vector<Span*>& spans, const Encoder& encoder);
+        virtual ~EncoderContext();
         iterator begin();
         iterator end();
         const_iterator begin() const;
         const_iterator end() const;
-        std::string ToString();
-        ::Span ToSpan();
-
-    private:
-        friend class Encoder<EncodingTypes::Thrift>;
-        void AddSpan(const ::Span& thriftSpan);
+        std::string ToString() const;
+        void AddSpan(const Span* span);
 
     private:
-        std::list<ContextElement> m_spans;
-        std::unique_ptr<apache::thrift::protocol::TBinaryProtocol> m_protocol;
-        boost::shared_ptr<apache::thrift::transport::TMemoryBuffer> m_buffer;
+        std::vector<ContextElement> m_spans;
+        const Encoder&              m_encoder;
```

Comment:
Since you are having reference, copy constructor, default constructor and copy assignment needs to be deleted, there is no true copy of the state.